src/Microsoft.ML.Core/Data/VBuffer.cs

```diff
@@ -8,6 +8,159 @@
 
 namespace Microsoft.ML.Runtime.Data
 {
+    /// <summary>
+    /// A readonly buffer that supports both dense and sparse representations. This is the
+    /// readonly representation type for all VectorType instances used as a guarantee that
+    /// the buffer cannot be modified.
+    /// </summary>
+    public readonly struct ReadOnlyVBuffer<T>
+    {
+        private readonly T[] _values;
+        private readonly int[] _indices;
+
+        /// <summary>
+        /// The logical length of the buffer.
+        /// </summary>
+        public readonly int Length;
+
+        /// <summary>
+        /// The number of items explicitly represented. This is == Length when the representation
+        /// is dense and &lt; Length when sparse.
+        /// </summary>
+        public readonly int Count;
+
+        /// <summary>
+        /// The values. Only the first Count of these are valid.
+        /// </summary>
+        public ReadOnlySpan<T> Values => _values;
```

Comment:
>Values => _values [](start = 31, length = 17)

I think this is a good beginning. I feel like though if we redesign `VBuffer`, my primary goal would be to get rid of all the difficulties that made my writing [this document](https://github.com/dotnet/machinelearning/blob/master/docs/code/VBufferCareFeeding.md) necessary.

That is, I think much of the stuff here should actually be in `VBuffer` *itself*.

In particular, I would move this to `VBuffer` and change this to `Values => _values.AsSpan(0, count)`, and `Indices => IsDense ? default : _indices.AsSpan(0, count)`.

We would need some separate object (I suspect internal) that is capable of mutating a `VBuffer`, to mutate a `VBuffer`.

In this world there would still I think be benefit to having a read-only version of `VBuffer`, since that thing could be a `ref struct` with direct fields of `ReadOnlySpan`.

*Extenally*, both `VBuffer` and `ReadOnlyVuffer` would both appear to be immutable objects, but `VBuffer`'s buffers would be internally reusable. `ROVB` though would be more efficient, since in that.